### ISSUE TITLE
Document how React treats different attributes

### ISSUE THREAD

**Issue by bvaughn** [reactjs/reactjs.org#80]
This issue was originally reported by @MMeent via facebook/react/issues/9220

Please refer to the original issue for the related discussion thread.

### Text of original issue:

As of #9106 there are quite a lot of tests for HTML's boolean properties on tags. According to these tests HTML properties should get rendered into empty strings or not set, e.g. `<div hidden={true} />` should render to `<div hidden="" />` and `<div hidden={false} />` should render to `<div />`. 

But when I look further, there are other tests which do similar things: `<a href={true} />` should render to `<a href="true" />`. This means that if I take `true` as my property value, I will not know what the result will be. 

The issue being: React currently does not have a clearly-defined and documented way with which they handle their props. Could the docs be updated to specify which type of prop will result in what? e.g. "any number gets parsed to a string (using Number.toString()), an object gets stringified by using `arg => Object.entries(arg).map(([key, val]) => `${key}: ${val};`).join()`, and booleans toggle the property."

In short, these are the bugs/undocumented features/contradictions that the tests expect to happen, plus my expectations:

- `prop={true}` gets parsed to [`prop=""`](https://github.com/facebook/react/blob/78761c387f7ab24b0dbbb63e9085b26b808e5330/src/renderers/dom/shared/__tests__/ReactDOMServerIntegration-test.js#L222) or [`prop="true"`](https://github.com/facebook/react/blob/78761c387f7ab24b0dbbb63e9085b26b808e5330/src/renderers/dom/shared/__tests__/ReactDOMServerIntegration-test.js#L209) for select properties  
    expected: either `prop=""` or `prop="prop"`, not both // as per checked=checked HTML standard
- `prop=""` gets parsed as [`prop=""`](https://github.com/facebook/react/blob/78761c387f7ab24b0dbbb63e9085b26b808e5330/src/renderers/dom/shared/__tests__/ReactDOMServerIntegration-test.js#L303) or [` `](https://github.com/facebook/react/blob/78761c387f7ab24b0dbbb63e9085b26b808e5330/src/renderers/dom/shared/__tests__/ReactDOMServerIntegration-test.js#L239)  
    expected: `prop=""`
- `prop="prop"` might get parsed as either [`prop="prop"`](https://github.com/facebook/react/blob/78761c387f7ab24b0dbbb63e9085b26b808e5330/src/renderers/dom/shared/__tests__/ReactDOMServerIntegration-test.js#L286), [`prop=""`](https://github.com/facebook/react/blob/78761c387f7ab24b0dbbb63e9085b26b808e5330/src/renderers/dom/shared/__tests__/ReactDOMServerIntegration-test.js#L231) or even [` `](https://github.com/facebook/react/blob/78761c387f7ab24b0dbbb63e9085b26b808e5330/src/renderers/dom/shared/__tests__/ReactDOMServerIntegration-test.js#L376)  
    expected: `prop="prop"`
- `prop="foo"` => [`prop=""`](https://github.com/facebook/react/blob/78761c387f7ab24b0dbbb63e9085b26b808e5330/src/renderers/dom/shared/__tests__/ReactDOMServerIntegration-test.js#L245) or [`prop="foo"`](https://github.com/facebook/react/blob/78761c387f7ab24b0dbbb63e9085b26b808e5330/src/renderers/dom/shared/__tests__/ReactDOMServerIntegration-test.js#L286)  
    expected: `prop="foo"`
- `prop={false}` => [` `](https://github.com/facebook/react/blob/78761c387f7ab24b0dbbb63e9085b26b808e5330/src/renderers/dom/shared/__tests__/ReactDOMServerIntegration-test.js#L282) or [`prop="false"`](https://github.com/facebook/react/blob/78761c387f7ab24b0dbbb63e9085b26b808e5330/src/renderers/dom/shared/__tests__/ReactDOMServerIntegration-test.js#L316)  
    expected: ` `
- `prop={['foo', 'bar']}` => [`prop=""`](https://github.com/facebook/react/blob/78761c387f7ab24b0dbbb63e9085b26b808e5330/src/renderers/dom/shared/__tests__/ReactDOMServerIntegration-test.js#L252) or `prop="foo,bar"` (hand-tested, no automatic test available)  
    expected: `prop="foo bar"`
- `prop={{foo: 'bar', baz: 5}}` => [`prop=""`](https://github.com/facebook/react/blob/78761c387f7ab24b0dbbb63e9085b26b808e5330/src/renderers/dom/shared/__tests__/ReactDOMServerIntegration-test.js#L257) or [`prop="foo: bar; baz: 5;"`](https://github.com/facebook/react/blob/master/src/renderers/dom/shared/__tests__/ReactDOMComponent-test.js#L1180) or `prop="[object Object]"` (hand-tested, no automatic test available)  
    expected: `prop="foo: bar; baz: 5"` // as per style object transform

Hacky showcase: https://jsfiddle.net/ox8a7vfe/3/

Are these features or are these bugs?

If they are bugs, and work is going to be done on the props value parsing, then the following might be a nice addition:

- `prop={{foo: true, bar: true, baz: false}}` => `prop="foo bar"` // render all `true` keys, disregard `false` keys, like classNames.

**Comment by mrscobbler** [comment]
It looks like this blog post: https://reactjs.org/blog/2017/09/08/dom-attributes-in-react-16.html addresses a lot of the concerns in this particular issue.

Do we want to create a new docs page that summarizes what the blog post talks about and reference the blog post?

**Comment by bvaughn** [comment]
Perhaps, or perhaps we could add to this page: https://reactjs.org/docs/dom-elements.html

To be honest, I moved this issue from the old repo over the weekend along with a lot of others. I didn't read too deeply into it.

Would you be interested in working on this @mrscobbler ?

**Comment by mrscobbler** [comment]
Sure! I can work on it. I think adding to the existing page would be great.

**Comment by bvaughn** [comment]
This issue is all yours! 😄

I've added an "in-progress" label so that others will know not to start work on the issue. If you change your mind about the issue, no worries! Just let me know so that I can remove the label and free it up for someone else to claim.

Cheers!

**Comment by bvaughn** [comment]
Hey @mrscobbler! Just wanting to confirm that this issue is still being worked on.

No hurry of course. I'm just going through in-progress issues to see if any have been dropped. 😄

**Comment by mrscobbler** [comment]
Yes! Still planning on working on it. I'll make sure to comment on this if for some reason I can't get to it.

**Comment by bvaughn** [comment]
Great! Thanks!

**Comment by Alexectramagneta** [comment]
Pedophile language protection of children

**Comment by Alexectramagneta** [comment]
Pedophile language protection of children

**Comment by ManitejChowdary1** [comment]
Hello i would like to contribute to this

**Comment by Alexectramagneta** [comment]
Ummm I really need to seek some legal advice I think if you would like to phone me to discuss I can give u my number

**Comment by suii2210** [comment]
Hi! I would like to work on this issue.

My plan:
- Add a new subsection to `src/content/learn/describing-the-ui.md`
  called "How React Handles HTML Attributes".
- Document the differences between HTML attributes and JSX props
  (className, htmlFor, boolean attributes, inline styles, data/aria attributes, unknown attributes).
- Include clear examples and align the tone with the existing docs.

Please let me know if this approach is okay, or if you prefer this to be added in a different page.

**Comment by suii2210** [comment]
> Hi! I would like to work on this issue.
> 
> My plan:
> 
> * Add a new subsection to `src/content/learn/describing-the-ui.md`
>   called "How React Handles HTML Attributes".
> * Document the differences between HTML attributes and JSX props
>   (className, htmlFor, boolean attributes, inline styles, data/aria attributes, unknown attributes).
> * Include clear examples and align the tone with the existing docs.
> 
> Please let me know if this approach is okay, or if you prefer this to be added in a different page.

Correction: The subsection will actually be added to 
`src/content/learn/writing-markup-with-jsx.md` instead of `describing-the-ui.md`,
since this page specifically covers JSX markup.

**Comment by vishpatel30** [comment]
Hi Would like to contribute to this issue

**Comment by wdev6666** [comment]
Hi @bvaughn  I’d like to work on this issue. Could you please assign it to me?

**Comment by Tusharv99** [comment]
want to start this , can anyone guide me how can i do contribution in github .
